Fix policy compatibility for well-known deployments such as kube-dns and etcd-operator

tgraf opened the pull request:
Make Cilium aware of kube-dns and the etcd-operator used for managed Kubernetes
instead of requiring the user to assign fixed identities. This guarantees that
existing label based kube-dns network policies continue working even in this
new deployment mode. Previously, kube-dns would change its security relevant
labels to reserved:kube-dns

Fixes: #5529
Fixes: #5791 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5844)
<!-- Reviewable:end -->

jrajahalme commented on Documentation/kubernetes/install/eks.rst:
```diff
@@ -54,21 +54,12 @@ Create EKS Cluster
        kubectl -n kube-system set env ds aws-node AWS_VPC_K8S_CNI_EXTERNALSNAT=true
 
-#. Assign a fixed security identity to ``kube-dns`` by  adding the label ``io.cilium.fixed-identity: kube-dns``
+#. Restart kube-dns to ensure that it is being managed by Cilium.
```
We should apply the same instruction in the minikube GSG as this step is necessary even there if any egress policies are applied?

tgraf replied:
The minikube GSG is not using the etcd operator yet. I think we are close to being confident enough to switch over.

jrajahalme commented on Documentation/policy/language.rst:
```diff
@@ -996,3 +996,20 @@ resources.
+contact any external dependencies. The purpose of this is to allow
+bootstrapping Cilium and enable network connectivity with policy enforcement in
+the cluster for essential services without depending on any dependencies.
+
```
The term "external dependencies" or "dependencies" is maybe unnecessarily vague here. Could it mean anything else than a kv-store?

ianvernon replied:
without any dependencies? "depending on any dependencies" is redundant.

jrajahalme commented on examples/kubernetes/addons/etcd-operator/README.md:
```diff
@@ -51,18 +51,11 @@ in the Kubernetes cluster, this can be achieved by running:
 
-Deploy kube-dns and make sure it contains the required label
-------------------------------------------------------------
+Restart kube-dns/coredn to ensure it is managed by Cilium
```
Spelling of "coredn"?

aanm commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +58,139 @@ const (
+	// ReservedEKSKubeDNS is the reserved identity used for kube-dns on EKS
+	ReservedEKSKubeDNS NumericIdentity = 103
+
+	// ReservedEKSCoreDNS is the reserved identity used for CoreDNS
```
this needs to be fixed ^

jrajahalme reviewed (approved):
Small nits only.

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -996,3 +996,21 @@ resources.
+Well-known Identities
+---------------------
+
+The following is a list of well known identities which Cilium is aware of
```
well-known

ianvernon commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +58,139 @@ const (
+		"k8s:k8s-app=kube-dns",
+		"k8s:eks.amazonaws.com/component=kube-dns",
+		fmt.Sprintf("k8s:%s=kube-system", api.PodNamespaceLabel),
+		fmt.Sprintf("k8s:%s=default", api.PolicyLabelServiceAccount),
```
The comment on line 168 says that the serviceaccount is kube-dns, not default. Should this label match the one in the comment?

tgraf replied:
Awesome catch. The comment was right, the code was wrong.

aanm commented on test/k8sT/manifests/v1.2/cilium-ds-patch.yaml:
```diff
@@ -17,8 +17,6 @@ spec:
         - "--k8s-require-ipv4-pod-cidr"
         - "--pprof=true"
         - "--log-system-load"
-        - "--fixed-identity-mapping=128=kv-store"
```
@tgraf Don't remove these lines

tgraf replied:
Why not?

tgraf replied:
Oh, because of the upgrade/downgrade.

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -145,6 +146,22 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
 
+		var oldLabels map[string]string
+		// Workaround to allow upgrading from 1.2 to newer Cilium versions
+		if strings.HasPrefix(oldVersion, "1.2") {
```
What happens with 1.0 and 1.1? Does not matter here?

tgraf replied:
1.0 and 1.1 do not have etcd operator support

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -145,6 +146,22 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
+			// Apply patch for fixed identities for if oldVersion is 1.2
+			oldLabels, err = patchKubeDNSWithFixedIdentity(kubectl)
+			ExpectWithOffset(1, err).To(BeNil(), "Unable to patch kube-dns with fixed identity labels")
+			defer func() {
```
I would this function always in the cleanup function, if not will not happen at all, if test fail, ginkgo call panic, and cleanup never happens. 

Please, remember, in cleanup if you do a assert, and fail, next execution will not work.

tgraf replied:
I'm not exactly sure how to do this and where

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -237,6 +254,13 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
 
+		// We need to revert the patch we have implemented in kube-dns for 1.2
+		// otherwise the well known identities introduced in 1.3 will not work.
+		if strings.HasPrefix(oldVersion, "1.2") {
```
The same as above, 1.1 and 1.0 does not matter here?

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -280,3 +304,53 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
+	oldLabelsMap := fb.KVOutput()
+
+	// Apply fixed identity patch
+	cmdRes := kubectl.Exec(fmt.Sprintf(`kubectl patch -n kube-system %s --type merge -p '{"spec":{"template":{"metadata":{"labels":{"io.cilium.fixed-identity":"kube-dns"}}}}}'`,
```
```
		cmdRes := kubectl.Exec(fmt.Sprintf(`%s patch -n %s %s --type merge -p '{"spec":{"template":{"metadata":{"labels":{"io.cilium.fixed-identity":"kube-dns"}}}}}'`, 
		helpers.KubectlCMD, helpers.KubeSystemNamespace
		dnsDeployment))
```

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -280,3 +304,53 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
+
+	b, err := json.Marshal(oldLabels)
+	if err != nil {
+		return err
```
`return fmt.Errorf("Cannot marshal old labels:", err)`

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -280,3 +304,53 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
+// revertPatchKubeDNSWithFixedIdentity reverts the changes made by
+// PatchKubeDNSWithFixedIdentity
+func revertPatchKubeDNSWithFixedIdentity(kubectl *helpers.Kubectl, oldLabels map[string]string) error {
+	dnsDeployment := "deployment/kube-dns"
```
I would move this, to something helpers.GetDNSDeploymentName() similar as `DNSDeployment()`
```
	dnsDeployment := "deployment/kube-dns"
	switch helpers.GetCurrentK8SEnv() {
	case "1.11", "1.12":
		dnsDeployment = "deployment/coredns"
	}
```

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -280,3 +304,53 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
+		return err
+	}
+
+	s := fmt.Sprintf(`kubectl patch -n kube-system %s --type json -p '[{"op": "replace","path": "/spec/template/metadata/labels","value": %s}]'`,
```
use:
helpers.kubectlCMD as kubectl. 
helpers.KubeSystemNamespace as kube-system

eloycoto commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +58,139 @@ const (
+	//   k8s:k8s-app=kube-dns
+	//   k8s:io.cilium.k8s.policy.cluster=default
+	wellKnown.add(ReservedCoreDNS, []string{
+		"k8s:k8s-app=kube-dns",
```
You repeated this line two times, maybe use a cons here?

tgraf replied:
I don't think it's a good idea to use constants here because these things may change and having a single constant for things like the kube-dns label can lead to blindly adjusting that constant that will then affect well-known identities. Once merged, we can never change these well-known identities anymore.

eloycoto commented on pkg/identity/numericidentity.go:
```diff
@@ -55,8 +58,139 @@ const (
+		"k8s:app=etcd",
+		"k8s:etcd_cluster=cilium-etcd",
+		"k8s:io.cilium/app=etcd-operator",
+		fmt.Sprintf("k8s:%s=kube-system", api.PodNamespaceLabel),
```
Maybe use here `helpers.KubeSystemNamespace` instead of "kube-system" is a good idea.

tgraf replied:
Same here, plus we definitely don't want to depend on test code from the regular identity package.

eloycoto commented on test/helpers/utils.go:
```diff
@@ -294,6 +294,17 @@ func WriteOrAppendToFile(filename string, data []byte, perm os.FileMode) error {
 
+// GetDNSDeploymentName returns the name of the deployment depending on the
+// deployed Kubernetes version
+func GetDNSDeploymentName() string {
```
@tgraf I think that you forget to add `k8st/Updates.go` changes are no longer here.

tgraf replied:
Leftover, it is no longer needed. Removed.

tgraf committed:
k8s: Make Cilium aware of well known deployments such as kube-dns and etcd-operator

Make Cilium aware of kube-dns and the etcd-operator used for managed Kubernetes
instead of requiring the user to assign fixed identities. This guarantees that
existing label based kube-dns network policies continue working even in this
new deployment mode. Previously, kube-dns would change its security relevant
labels to reserved:kube-dns

Fixes: #5529
Fixes: #5791

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: André Martins <andre@cilium.io>

aanm commented on test/k8sT/Updates.go:
```diff
@@ -140,10 +140,15 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
-		By("Deploying etcd-operator")
-		err = kubectl.DeployETCDOperator()
-		Expect(err).To(BeNil(), "Unable to deploy etcd operator")
+		switch oldVersion {
```
@tgraf This doesn't matter, etcd-operator is just a pod like all the others, the fact it's not used in `1.0-1.2` should be irrelevant for the upgrade test. This will make the upgrade test to fail as, when upgrading to 1.3/master, there isn't a etcd cluster available so 1.3 will never be ready.

tgraf replied:
Yes, I want to test that v1.2 comes up clean first and then involve the ETCD operator again.
